qiskit/transpiler/passes/mapping/algorithm/flexlayer_heuristics.py

```diff
@@ -0,0 +1,433 @@
+                             for e in ece.cand_edges]
+
+                    min_cost, edge = min(zip(costs, ece.cand_edges))
+
+                    edge = self._fix_swap_direction(edge)
+                    new_dag.apply_operation_back(SwapGate(), [qreg[edge[0]], qreg[edge[1]]])
+                    layout.swap(edge[0], edge[1])
+
+                    logger.debug("%d-th inner iter. add a swap (%d, %d)", kin, edge[0], edge[1])
+                    logger.debug("resolved layout = %s", pprint.pformat(sorted(layout.items())))
+
+                    dones = self._find_done_gates(blocking_gates, layout=layout)
+                    if dones:
+                        break
+
+                if kin == max_n_inner_loops:
+                    raise TranspilerError("Unknown error (maybe infinite inner loop)")  # bug
+
+            if k == max_n_iteration:
+                raise TranspilerError("Unknown error: #iteration reached max_n_iteration")
+
+        return new_dag, layout
+
+    def _next_blocking_gates_from(self, blocking_gates, layout):
+        """next blocking gates from blocking_gates for layout with finding applicable gates (dones)
+        """
+        leadings = set(blocking_gates)  # new leading gates
+        dones = []
+        new_dones = self._find_done_gates(leadings, layout)
+        while new_dones:
+            dones.extend(new_dones)
+            leadings = self._update_leading_gates(leadings, new_dones)
+            new_dones = self._find_done_gates(leadings, layout)
+
+        assert len(dones) == len(set(dones))
+        # dones must be list (order is essential!)
+
+        return frozenset(leadings), dones
+
+    def _find_done_gates(self, blocking_gates, layout):
+        dones = []
+        for n in blocking_gates:
+            qargs = self._dg.qargs(n)
+            if self._dg.gate_name(n) == "barrier":
+                dones.append(n)
+            elif len(qargs) == 1:  # 1-qubit gate (including measure)
+                dones.append(n)
+            elif len(qargs) == 2:  # CNOT(2-qubit gate)
+                dist = self._coupling.distance(layout[qargs[0]],
+                                               layout[qargs[1]])
+                if dist == 1:
+                    dones.append(n)
+            else:
+                raise TranspilerError("DG contains unknown >2 qubit gates")
+
+        return dones
+
+    def ancestors(self, n: int) -> set:
+        """
+        Ancestor gates of the gate `n`
+        Args:
+            n: Index of the gate in the `self._graph`
+        Returns:
+            Set of indices of the ancestor gates.
+        """
+        return self._ancestors.ancestors(n)
+
+    def _update_leading_gates(self, leading_gates, dones):
+        news = set(leading_gates)
+        for n in dones:
+            news |= set(self._dg.gr_successors(n))
+
+        news -= set(dones)
+
+        rmlist = []
+        for n in news:
+            if news & self.ancestors(n):
```

Comment:
I don't feel strongly about this suggestion, so ignore it if you want. This is a private method which already operates on `self._dg`, so you could replace this with `self._dg.ancestors(n)` (or `self._ancestors.ancestors(n)` if you don't make the other change).

My next question is, if I have an instance of `FlexlayerHeuristics`, does it make sense for that instance to have a (public?) `ancestors` method? I am certainly not very familiar with what a `Heuristic` instance should be able to do. Perhaps you can give me an example of a context that you'd want the ancestors of a gate at this level of abstraction?